controllers/workflow_controller.go

```diff
@@ -0,0 +1,454 @@
+		return
+	}
+
+	if _, ok := old[addonapiv1.AddonCompleteLabel]; ok {
+		//c.logger.Infof("mergeLabels remove complete label.")
+		delete(old, addonapiv1.AddonCompleteLabel)
+	}
+	for k, v := range old {
+		merged[k] = v
+	}
+}
+
+// add or remove addon finalizer according to new instance
+func (c *wfreconcile) mergeFinalizer(old, new []string) []string {
+	addFinalize := false
+	for _, f := range new {
+		if f == addonapiv1.FinalizerName {
+			// should add finalizer
+			addFinalize = true
+			break
+		}
+	}
+
+	if addFinalize {
+		needappend := true
+		for _, f := range old {
+			if f == addonapiv1.FinalizerName {
+				needappend = false
+			}
+		}
+		if needappend {
+			old = append(old, addonapiv1.FinalizerName)
+		}
+		//c.logger.Infof("mergeFinalizer after append addon finalizer %#v", old)
+		return old
+	}
+
+	// otherwise, remove finalizer
+	ret := []string{}
+	for _, f := range old {
+		if f == addonapiv1.FinalizerName {
+			continue
+		}
+		ret = append(ret, f)
+	}
+	//c.logger.Infof("mergeFinalizer after remove addon finalizer %#v", ret)
+	return ret
+}
+
+func (c *wfreconcile) mergeResources(res1, res2 []addonv1.ObjectStatus) []addonv1.ObjectStatus {
+	merged := []addonv1.ObjectStatus{}
+	check := make(map[string]addonv1.ObjectStatus)
+	mix := append(res1, res2...)
+	for _, obj := range mix {
+		id := fmt.Sprintf("%s-%s-%s", strings.TrimSpace(obj.Name), strings.TrimSpace(obj.Kind), strings.TrimSpace(obj.Group))
+		check[id] = obj
+	}
+	for _, obj := range check {
+		merged = append(merged, obj)
+	}
+	return merged
+}
+
+func (c *wfreconcile) removeFinalizer(addon *addonv1.Addon) {
+	if common.ContainsString(addon.ObjectMeta.Finalizers, addonapiv1.FinalizerName) {
+		addon.ObjectMeta.Finalizers = common.RemoveString(addon.ObjectMeta.Finalizers, addonapiv1.FinalizerName)
+	}
+}
+
+func (c *wfreconcile) removeFromCache(addonName string) {
+	// Remove version from cache
+	if ok, v := c.versionCache.HasVersionName(addonName); ok {
+		c.versionCache.RemoveVersion(v.PkgName, v.PkgVersion)
+	}
+}
+
+// update addon meta ojbect first, then update status
+func (c *wfreconcile) updateAddon(ctx context.Context, updated *addonv1.Addon) (*addonv1.Addon, error) {
+	var errs []error
+
+	latest := &addonv1.Addon{}
+	err := c.client.Get(ctx, types.NamespacedName{Namespace: updated.Namespace, Name: updated.Name}, latest)
+	if err != nil || latest == nil {
+		return nil, err
+	} else {
+
+		if reflect.DeepEqual(updated, latest) {
+			c.log.Info(fmt.Sprintf("[updateAddon] latest and updated %s/%s is the same, skip", updated.Namespace, updated.Name))
+			return nil, nil
+
+		}
+
+		// update object metata only
+		updating := latest.DeepCopy()
+		updating.Finalizers = c.mergeFinalizer(latest.Finalizers, updated.Finalizers)
+		updating.ObjectMeta.Labels = map[string]string{}
+		c.mergeLabels(latest.GetLabels(), updated.GetLabels(), updating.ObjectMeta.Labels)
+
+		err := c.client.Status().Update(ctx, updating, &client.UpdateOptions{})
```

Comment:
The logic for updating addon status needs to be shared across controllers, it's not maintainable here, we need to add shared functions under package `addons`.

Reply:
hmm... not quite understand your point, the client.Client is provided by the controller-runtime, and can be used anywhere in the repo.

Reply:
9f5e34b986ed223f8aa974644b1a6e58707a6e85